File: BinaryTreePreorderTraversal/BinaryTreePreorderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        tmp = []
        if root is None:
            return res
        while True:
            while root:
                res.append(root.val)
                tmp.append(root)
                root = root.left
            if len(tmp) == 0:
                return res    
            node = tmp.pop()
            root = node.right
#         res = []
#         dfs_preorder(root, res)
#         return res

# def dfs_preorder(root, res):
#     if root is None:
#         return
#     res += root.val,
#     if root.left:
#         dfs_preorder(root.left, res)
#     if root.right:
#         dfs_preorder(root.right, res)

File: BinaryTreePreorderTraversal/BinaryTreePreorderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        stack<TreeNode*> tmp;
		vector<int> result;
		if (root == nullptr)
			return result;
		while (true) {
			while (root != nullptr) {
				result.push_back(root->val);
				tmp.push(root);
				root = root->left;
			}
			if (tmp.size() == 0)
				return result;
			TreeNode* node = tmp.top();
			tmp.pop();
			root = node->right;
		}
    }
};